compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/calls/Arguments.kt

```diff
@@ -375,20 +376,43 @@ private fun Candidate.getExpectedTypeWithSAMConversion(
     val firFunction = symbol.fir as? FirFunction<*> ?: return null
     if (!context.bodyResolveComponents.samResolver.shouldRunSamConversionForFunction(firFunction)) return null
 
-    if (!argument.isFunctional(session)) return null
-
     // TODO: resolvedCall.registerArgumentWithSamConversion(argument, SamConversionDescription(convertedTypeByOriginal, convertedTypeByCandidate!!))
 
-    return context.bodyResolveComponents.samResolver.getFunctionTypeForPossibleSamType(candidateExpectedType).apply {
-        usesSAM = true
-    }
+    val expectedFunctionType = context.bodyResolveComponents.samResolver.getFunctionTypeForPossibleSamType(candidateExpectedType)
+    return if (argument.isFunctional(session, scopeSession, expectedFunctionType)) {
```

Comment:
Please use `runIf(cond) { block }` instead of `if (cond) { block } else null`

Reply:
Good to learn that! Changed as per the suggestion.